commit 67a2471437e79487a269ad2c9b542fbe8e6d41dd
Author: Miss Islington (bot) <31488909+miss-islington@users.noreply.github.com>
Date:   Wed Jul 15 05:36:36 2020 -0700

    00351: Avoid infinite loop in the tarfile module
    
    Avoid infinite loop when reading specially crafted TAR files using the tarfile module
    (CVE-2019-20907).

diff --git a/Lib/tarfile.py b/Lib/tarfile.py
index 5f1a979ad0..827a4db27c 100755
--- a/Lib/tarfile.py
+++ b/Lib/tarfile.py
@@ -1226,6 +1226,8 @@ class TarInfo(object):
 
             length, keyword = match.groups()
             length = int(length)
+            if length == 0:
+                raise InvalidHeaderError("invalid header")
             value = buf[match.end(2) + 1:match.start(1) + length - 1]
 
             # Normally, we could just use "utf-8" as the encoding and "strict"
diff --git a/Lib/test/recursion.tar b/Lib/test/recursion.tar
new file mode 100644
index 0000000000..b823725196
Binary files /dev/null and b/Lib/test/recursion.tar differ
diff --git a/Lib/test/test_tarfile.py b/Lib/test/test_tarfile.py
index 3091ce77cb..241e42b814 100644
--- a/Lib/test/test_tarfile.py
+++ b/Lib/test/test_tarfile.py
@@ -373,6 +373,13 @@ class CommonReadTest(ReadTest):
                 with self.assertRaisesRegex(tarfile.ReadError, "unexpected end of data"):
                     tar.extractfile(t).read()
 
+    def test_length_zero_header(self):
+        # bpo-39017 (CVE-2019-20907): reading a zero-length header should fail
+        # with an exception
+        with self.assertRaisesRegex(tarfile.ReadError, "file could not be opened successfully"):
+            with tarfile.open(support.findfile('recursion.tar')) as tar:
+                pass
+
 class MiscReadTestBase(CommonReadTest):
     def requires_name_attribute(self):
         pass
diff --git a/Misc/NEWS.d/next/Library/2020-07-12-22-16-58.bpo-39017.x3Cg-9.rst b/Misc/NEWS.d/next/Library/2020-07-12-22-16-58.bpo-39017.x3Cg-9.rst
new file mode 100644
index 0000000000..ad26676f8b
--- /dev/null
+++ b/Misc/NEWS.d/next/Library/2020-07-12-22-16-58.bpo-39017.x3Cg-9.rst
@@ -0,0 +1 @@
+Avoid infinite loop when reading specially crafted TAR files using the tarfile module (CVE-2019-20907).
